### ISSUE TITLE
ASB 07/2017

### ISSUE THREAD

**Issue by MSe1969** [derfelot/discussion#10]
The Google ASB 07/2017 is out. (Not yet with links to the code)
Looks like a lot of work this time, again "Media" being the 'hot spot' . . .

**Comment by MSe1969** [comment]
As for release planning: All-in-one? 
- the app features
- 3.4.113
- CVE-2016-5349
- ASB 07/2017

Or should we pack the app features and 3.4.113 and CVE-2016-5349 (which seems to run stable) into an 'Interim-Release' and take our time to do the ASB (1x runtime, 6x framework, 1x library, 9x media, 1x system UI, 8x upstream-kernel (if applicable), 17x CAF kernel (if applicable)?

**Comment by Micha-Btz** [comment]
I would say all in one. Normally we make testbuilds after patching CVE's from ASB 07/2017 for us and so there are enough test's I think. And the kernel is tested about several weeks. And at least it is more work do more than one release :-)

**Comment by Micha-Btz** [comment]
Links to the patches now available. I will start with the kernel fixes. Which one you will fix @MSe1969 ?
As far as i know @derfelot is totally busy. Just that we not work on the same things.

**Comment by MSe1969** [comment]
Yes @Micha-Btz - if you could do the kernel stuff, that would be very good.
I am right now working on the other ("non-kernel") patches for LOS13, and will have from that exercise then a lot of patch files available, which I will try to reuse as much as possible for our ROM. So I will do the other patches.

**Comment by MSe1969** [comment]
I've stored a tracking file for the July ABS into this repo, please add also the kernel related stuff - will start now the patching for the non-kernel stuff

**Comment by Micha-Btz** [comment]
Kernel components all fiexd. Some we have fixed before the ASB.
CVE-2017-6074 l | EoP | High | Networking subsystem
CVE-2017-5970  | DoS | High | Networking subsystem
CVE-2015-5707 | EoP | Moderate | SCSI driver
CVE-2017-0710  EoP | Moderate | TCB
CVE-2017-7308 | EoP | Moderate | Networking driver
CVE-2014-9731File system

**Comment by Micha-Btz** [comment]
Qualcomm components
I exclude bootloader an d camera fixes.
- [ ] CVE-2017-8262 GPU driver
- [ ] CVE-2017-8263 Anonymous shared memory subsystem
- [ ] CVE-2017-8267 Anonymous shared memory subsystem
- [ ] CVE-2016-5863 USB HID driver
- [ ] CVE-2017-8243 SoC driver
- [ ] CVE-2017-8246 Sound driver
- [ ] CVE-2017-8256 Wi-Fi driver
- [ ] CVE-2017-8257 Video driver
- [ ] CVE-2017-8259 SoC driver
- [ ] CVE-2017-8265 Video driver
- [ ] CVE-2017-8266 Video driver
- [ ] CVE-2017-8270 Wi-Fi driver
- [ ] CVE-2017-8271 Video driver
- [ ] CVE-2017-8272 Video driver
- [ ] CVE-2017-8254 Sound driver
- [ ] CVE-2017-8269 IPA Driver

**Comment by MSe1969** [comment]
Btw, CVE-2017-0670 patch (bionic) screws up everything in cm-13.0, so trying to find a solution there first (it'll definitely also screw up cm-12.1 then)

**Comment by Micha-Btz** [comment]
I have had some trouble with two patches so I reverted these. I have updated the list above but I think we are ready to go.

**Comment by MSe1969** [comment]
great - I've implemented all patches except CVE-2017-0670, where a cherry-pick/backport of another change is necessary first before continuing. Test build of that cherry-pick running, afterwards, will continue.

**Comment by Micha-Btz** [comment]
@MSe1969 Will you please update the local manifest since I have no overview with packages you have forked or which branches you have created.

**Comment by MSe1969** [comment]
Now, CVE-2017-0670 has also been implemented. The local manifest has been updated (cm-12.1-test). For the July ASB, there was no new fork necessary. So the situation is as follows:
- Kernel has been patched by @Micha-Btz and @hahnjo (cm-12.1-test)
- My replacements f. SetupWizard, LockClock and Trebuchet are forks in my repository with branch cm-12.1 (as they are new, there is no need for a cm-12.1-test branch)
- All relevant ASB patches have been implemented (status file updated) into the respective cm-12.1-test branches
- local manifest is updated in branch cm-12.1-test

**However** - external/libhevc doesn't compile (syntax error), there seems to be an upstream commit missing. I don't have time now (most probably only on Friday) to look after that, maybe one of you can investigate?

**Comment by Micha-Btz** [comment]
@MSe1969 I have no compile problem. Rom build and run fine.

**Comment by Micha-Btz** [comment]
I have give it a try yesterday but you are right there is something strange in the code.
Maybe we should merge https://android.googlesource.com/platform/external/libhevc/+/android-5.1.1_r24
this tag and give the patch a new chance. For build test I have reverted the relevant patch.

But there is also a problem in Download Provider
https://gist.github.com/Micha-Btz/a5ca6d960e4b3369989b7f3c3ea71588

**Comment by MSe1969** [comment]
Unfortunately, I won't have any time today and most probably also not tomorrow. If you can take care, please go ahead. One comment: Our manifest is on AOSP tag 5.1.1_r38 - so shouldn't we merge this tag, if our forks are far behind? (Even if there hasn't been any change since)

**Comment by Micha-Btz** [comment]
I will make a test, if we could use the cm-13 tree. Download Provider the only thing I can do is to revert the patch. I simply don't understand java.

**Comment by hahnjo** [comment]
I will try to have a look at the two problems now ;-)

**Comment by hahnjo** [comment]
I've fixed DownloadProvider and uploaded the fix to my fork: https://github.com/hahnjo/android_packages_providers_DownloadProvider

With that, the build has succeeded, will test in a minute

**Comment by Micha-Btz** [comment]
@hahnjo  @MSe1969 
The download provider patch works :-)
To take external/libhevc from the cm-13.0 tree doesn't work there are to massive changes so that it doesn't compile. But it is really strange seems that libhevc from CyanogenMoD/LOS differs really from android source. In my opinion there is missing a complete merge.
here the diff from it 
[0001-Added-SEI-structures-and-reordered-ihevc_structs.h-c.patch.zip](https://github.com/derfelot/discussion/files/1147748/0001-Added-SEI-structures-and-reordered-ihevc_structs.h-c.patch.zip)

maybe we switch to the original repo or we just ignore the patch.

**Comment by MSe1969** [comment]
Thanks @hahnjo for finding the issue. I've amended my commit accordingly. so it appears as one commit, as it should be and not as several commits.
I'll have a look now to libhevc

**Comment by MSe1969** [comment]
Just a hint, forgot to patch bionic - have now first applied a patch, which was also necessary in cm-13.0 to prevent the "overkill crashing" which has happened when applying the ASB patch. That ASB patch however cannot be cherry-picked, I need to apply it manually.

**Comment by MSe1969** [comment]
hmm... - looking at the bionic patch, the coding really is significantly different, not really sure how we would backport that change . . .

**Comment by MSe1969** [comment]
I have fixed libhevc (in fact, only one '#define' statement needed to be backported), have incorporated/amended the DownloadProvider fix from hahnjo and have set the version string.

But the bionic change needs to be deferred, the coding is significantly different, if you also look at the main logic and the call chains. I suggest to ship w/o bionic and work on bionic separately.

The test build looks good so far - please create your own test builds and advise, whether you're OK with shipping.

Besides finally fixing bionic, we can also separately look to merge the more recent 5.1.1_r38 tag into our forks, where applicable. OK?

**Comment by hahnjo** [comment]
5.1.1_r38 will probably include the security fixes that we backported by hand, no?

**Comment by MSe1969** [comment]
No, 5.1.1_r38 is the last tag published by Google for "Lollipop" - it has been published in July 2016 and corresponds to security string 2016-07-05 - so it's one year old.
The CM forks have sometimes been forked way before, so at least merging the "non-security related" stuff (the sec.related hopefully is in through the ASB's) may improve the code base or may make it in future easier for us to backport ASB stuff. 
@Micha-Btz came with that topic with regards to libhevc - I do not see a **general** value here, because the LP code is still old, but in individual cases, this may make sense.

Regarding bionic, please have a closer look - the entire logic of unlinking the libs has entirely changed between LP and later android versions. The LP code calls the soinfo_unload function recursively and the unload happens through a si->CallDestructor() method (I have no access to the code now, so the names may be slightly different, but you'll understand), which is implmented also in linker.cpp, and from there, further calls into local methods occur. All of that is significantly different in MM or N, so I don't see, how we could backport the ASB change, without more or less backporting the entire new logic, which I don't really understand at all - and bionic is one of the core functionalities in android.
Any ideas?

**Comment by hahnjo** [comment]
@MSe1969 I think we should keep the security backports on a best-effort basis. However, the commit message is quite unclear and I don't really understand how important this change is...

**Comment by Micha-Btz** [comment]
The problem is when we not understand the commit and no one from us can backport the whole 
bionic and libhevc than I would say leave it out. Maybe we take more damage when we backport this wrong.

Other question:
Do you use encryption? On xda there is a message that initial encryption not work.

**Comment by hahnjo** [comment]
@Micha-Btz I have seen that. I will try to find some time for testing. Might be worth a try to encrypt with a clean install...

**Comment by hahnjo** [comment]
@MSe1969 you'd need to force push to DownloadsProvider. At the moment, there are two versions of the commit and a merge afterwards.

**Comment by MSe1969** [comment]
@hahnjo - just to make sure: What would need to be done exactly? I've invited you as a collaborator, so if you can quickly do it, appreciated :)

**Comment by MSe1969** [comment]
Coming back to the ASB topic, @hahnjo and @Micha-Btz - I have fixed the libhevc issue, so libhevc is done (done a test build and updated cm-12.1-test, running stable). However - bionic still open.

This is my opinion:
1. We should ship the ASB and defer this, but not abandon it
2. The described issue of unlinking seems to me a "chain" of issues working together before it actually becomes a vulnerability, so I would not see this as a vulnerability on its own, but still something which should be fixed - or we could ask the question, whether this would even apply. The ASB patch in fact "fine tunes" a check, which does not exist in our version at all...
3. How to tackle it? Actually, I don't see a good way to try patching the old code, so indeed backporting a significant chunk of the new coding logic seems to be the only way, but that sounds quite "tricky" to me. We could also try to simply send a polite e-mail to the developer at Google and ask him to provide us with a patch file (and hope that he even reacts)

**Comment by hahnjo** [comment]
@MSe1969 done ;-)

```
$ git reset --hard 539886b
$ git push --force push cm-12.1:cm-12.1
```

I agree, we should skip bionic for now. Please give me some time to figure out a solution for #11 before shipping.

**Comment by MSe1969** [comment]
Thanks :)

**Comment by hahnjo** [comment]
I've fixed #11 by reverting the commits for CVE-2016-5349 and opened #12 as a follow-up.

@MSe1969 @Micha-Btz IMO we are ready to release. Shall I do a clean build this evening and announce the new version?

**Comment by MSe1969** [comment]
Hi @hahnjo - thanks a lot! 

Regarding release/shipment, I suggest the following:

1. I would like to edit the announcement also in the project itself, as I also did some changes to a couple of apps (see my other issue), so I would like to take care this time.

2. I just did a clean build yesterday (microG patched), so re-doing it with the updated kernel won't take long. If you in parallel create a clean build (w/o microG patch), and upload it to AFH, I'll be happy to refer to it when I update the OP. 
 
3. @hahnjo - do you currently have edit rights in XDA on our project? The DB allows to explicitly enter a "contributor 1" and "contributor 2" and "additional contributors" - currently, you are listed as additional contributor. Could you please try? If you don't have edit rights, I suggest, for the time being, to switch you with @derfelot , as he is currently very busy work-wise and has no time to be around here...
Please feedback.

4. @Micha-Btz - I was not able so far to use the Download section of the project, that's why I always link to AFH. I don't receive any error, I simply get an incomplete PopUp asking me to pick a file, but that popup otherwise is entirely empty :-1:  - does it work for you?

Thanks & regards - M.

**Comment by MSe1969** [comment]
I've created #13 to follow up on the bionic patch

**Comment by hahnjo** [comment]
@MSe1969 looks like I can't edit OP, but my name also isn't "linked" as the rest...

Ok, can do the clean build, no problem ;-)

**Comment by MSe1969** [comment]
@hahnjo - I've switched you with @derfelot , you should now have edit rights

**Comment by MSe1969** [comment]
with the build now shipped and follow-on issues created, I'll close that one - will merge the branches accordingly
